### ISSUE TITLE
awlsimhw_linuxcnc: Changing linuxcnc I/O number cause error.

### ISSUE THREAD

**Issue by uncle-yura** [mbuesch/awlsim#20]
When in gui you change the number of inputs or outputs, the hal component is not updated.

```
Traceback (most recent call last):
  File "/home/lcnc/awlsim/examples/linuxcnc-demo/../..//awlsim-linuxcnc-hal", line 441, in <module>
    sys.exit(main())
  File "/home/lcnc/awlsim/examples/linuxcnc-demo/../..//awlsim-linuxcnc-hal", line 397, in main
    server.run()
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1683, in run
    self.__handleCommunicationBlocking()
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1506, in __handleCommunicationBlocking
    self.__handleCommunication()
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1466, in __handleCommunication
    self.__handleSocketComm(rlist)
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1439, in __handleSocketComm
    self.__handleClientComm(self.__sock2client[sock])
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1423, in __handleClientComm
    handler(self, client, msg)
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 1049, in __rx_RUNSTATE
    self.setRunState(self.STATE_RUN)
  File "/home/lcnc/awlsim/awlsim/coreserver/server.py", line 642, in setRunState
    raise e
AttributeError: Pin 'output.u31.6.active' does not exist
```
To reload hal you can use a bash script like this:
```
halcmd unloadusr awlsim
halcmd loadusr -Wn awlsim awlsim-linuxcnc-hal --listen localhost:4151 linuxcnc-demo.awlpro
```

But "unloadusr" breaks all networks, so it's better to call "source":
```
halcmd unloadusr awlsim
halcmd source awlsim.hal
```

It would also be nice to have this code called when the awlsim component crash:
```
        import linuxcnc
	command = linuxcnc.command()
	command.state( linuxcnc.STATE_OFF )
        command.state( linuxcnc.STATE_ESTOP )
	command.wait_complete()
```

For example, if a component stops during a zero search on a machine, this will result in physical damage to the axis, so it is better to stop the component. You can also add a parameter for the user what to do during an accident.

**Comment by mbuesch** [comment]
Thanks a lot for your bug report.

Yes, indeed. Changing the I/O configuration online is not supported by the linuxcnc hardware module, yet.
It should probably not crash, though.

**Comment by mbuesch** [comment]
I would rather let the user decide whether to restart Hal/LinuxCNC.
Resizing I/O areas is such a big change. It has other implications like the need for new nets in hal config.

So I fixed the crash and added a warning instead.

The E-Stop change will be tracked in a separate issue.

**Comment by uncle-yura** [comment]
If you enable the component to be updated, it will be possible to automatically connect the networks to the newly created pins. 

> 6.3. connect
> Connect a pin to a signal.
> example:
> hal.connect("pinname","signal_name")
> 
> 6.4. new_signal
> Create a New signal of the type specified.
> example"
> hal.new_sig("signalname",hal.HAL_BIT)

Just add a field in the GUI to enter the name of the automatically connected network.

**Comment by mbuesch** [comment]
Thanks for your suggestions.

I'd rather like to not do this.
This is due to a couple of reasons.
* Separation of concerns. The Awlsim LinuxCNC module knows nothing about signals. It only knows about its input and output pins. Connections are made in the .hal configuration file.
* Moving the signals (and their connections) to the Awlsim project would deviate from the basic concept that (all?) other LinuxCNC modules use. The HAL file is the place to make these connections.
* I don't think basic reconfiguration of the HAL I/O signals is a thing that has to be done often. It's Ok to demand a restart of the LinuxCNC system.

**Comment by uncle-yura** [comment]
In general, you are right, but the python hal component makes it possible to create networks, which means that this is not a deviation.

When working with hal files, you also do not know about networks, a typo in the name creates two different networks. I understand that LinuxCNC gives a lot of tuning flexibility, but I don’t like working in HAL, this process takes a lot of time, the program is very long, and debugging and navigation in it is very difficult.

I agree that reloading a component has an ugly solution, but when debugging a program on a machine, it is not always possible to constantly reload it. As an example, in one of my machines, when starting up, grease is turned on and at each load it poured a ton of oil onto the guides. Also, when you restart, you have to go to zero, which takes time.

How can it be a compromise that you can generate a hal file template with pins and networks initialized in awlpro?

**Comment by mbuesch** [comment]
I understand that it would be convenient not to restart.
But I still don't see how I could solve that in any way.
I cannot re-implement LinuxCNC's HAL configuration mechanism in Awlsim. That is the wrong place.
I agree that working with HAL configurations is not very nice. But if there's a need for an easier (e.g. graphical) HAL configuration, that has to be done in the LinuxCNC project.

Restarting of the HAL should not be triggered by Awlsim. There's no way Awlsim can know whether there are any side effects in other parts of the setup.
The user must decide whether to restart the HAL or not. The user is free to issue the halcmd manually.

**Comment by uncle-yura** [comment]
OK, you're right. What I describe only partially solves the problem. And I just thought that it would be nice to connect standard pins like "axis" or "motion". In practice, you will still have to connect external inputs and outputs. Anyway, you have to create a network and enter the physical input and output.

**Comment by mbuesch** [comment]
Fixed in Awlsim 0.72.0.
